Hey Stuart @stuartwdouglas, 
it is MP Config TCK now:
```

	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.io.NotSerializableException: io.quarkus.vertx.http.runtime.HttpHostConfigSource
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1185)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1553)
``` 
The new [config source](https://github.com/quarkusio/quarkus/pull/13223/files#diff-e99cad820a5d635678fd94a2cc4e4811f6247f3bce237c567052a15b3e516c25) looks fine

CC @radcortez

--- a/core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusDevModeLauncher.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/dev/QuarkusDevModeLauncher.java
@@ -230,6 +230,11 @@ public abstract class QuarkusDevModeLauncher {
             return (B) this;
         }
 
+        public B debugHost(String host) {
+            debugHost = host;
+            return (B) this;
+        }
+
         @SuppressWarnings("unchecked")
         public R build() throws Exception {
             prepare();
@@ -241,6 +246,7 @@ public abstract class QuarkusDevModeLauncher {
     private String debug;
     private Boolean debugPortOk;
     private String suspend;
+    private String debugHost = "localhost";
     private File projectDir;
     private File buildDir;
     private File outputDir;
@@ -310,14 +316,14 @@ public abstract class QuarkusDevModeLauncher {
             }
             if (debugPortOk) {
                 args.add("-Xdebug");
-                args.add("-Xrunjdwp:transport=dt_socket,address=0.0.0.0:5005,server=y,suspend=" + suspend);
+                args.add("-Xrunjdwp:transport=dt_socket,address=" + debugHost + ":5005,server=y,suspend=" + suspend);
             }
         } else if (debug.toLowerCase().equals("client")) {
             args.add("-Xdebug");
-            args.add("-Xrunjdwp:transport=dt_socket,address=localhost:5005,server=n,suspend=" + suspend);
+            args.add("-Xrunjdwp:transport=dt_socket,address=" + debugHost + ":5005,server=n,suspend=" + suspend);
         } else if (debug.toLowerCase().equals("true")) {
             args.add("-Xdebug");
-            args.add("-Xrunjdwp:transport=dt_socket,address=0.0.0.0:5005,server=y,suspend=" + suspend);
+            args.add("-Xrunjdwp:transport=dt_socket,address=" + debugHost + ":5005,server=y,suspend=" + suspend);
         } else if (!debug.toLowerCase().equals("false")) {
             try {
                 int port = Integer.parseInt(debug);
@@ -325,7 +331,7 @@ public abstract class QuarkusDevModeLauncher {
                     throw new Exception("The specified debug port must be greater than 0");
                 }
                 args.add("-Xdebug");
-                args.add("-Xrunjdwp:transport=dt_socket,address=0.0.0.0:" + port + ",server=y,suspend=" + suspend);
+                args.add("-Xrunjdwp:transport=dt_socket,address=" + debugHost + ":" + port + ",server=y,suspend=" + suspend);
             } catch (NumberFormatException e) {
                 throw new Exception(
                         "Invalid value for debug parameter: " + debug + " must be true|false|client|{port}");

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java
@@ -198,6 +198,7 @@ public class QuarkusDev extends QuarkusTask {
                 .buildDir(getBuildDir())
                 .outputDir(getBuildDir())
                 .debug(System.getProperty("debug"))
+                .debug(System.getProperty("debugHost"))
                 .suspend(System.getProperty("suspend"));
 
         if (getJvmArgs() != null) {

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -186,6 +186,9 @@ public class DevMojo extends AbstractMojo {
     @Parameter(defaultValue = "${suspend}")
     private String suspend;
 
+    @Parameter(defaultValue = "localhost")
+    private String debugHost;
+
     @Parameter(defaultValue = "${project.build.directory}")
     private File buildDir;
 
@@ -582,6 +585,7 @@ public class DevMojo extends AbstractMojo {
                 .outputDir(outputDirectory)
                 .suspend(suspend)
                 .debug(debug)
+                .debugHost(debugHost)
                 .debugPortOk(debugPortOk)
                 .deleteDevJar(deleteDevJar);
 

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java
@@ -10,6 +10,7 @@ import java.security.CodeSource;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
+import java.util.OptionalInt;
 import java.util.logging.Level;
 import java.util.stream.Collectors;
 
@@ -28,6 +29,7 @@ import io.quarkus.deployment.builditem.ApplicationStartBuildItem;
 import io.quarkus.deployment.builditem.ExecutorBuildItem;
 import io.quarkus.deployment.builditem.LaunchModeBuildItem;
 import io.quarkus.deployment.builditem.LogCategoryBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigurationSourceBuildItem;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
 import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
 import io.quarkus.deployment.builditem.ShutdownListenerBuildItem;
@@ -47,6 +49,7 @@ import io.quarkus.vertx.http.deployment.devmode.HttpRemoteDevClientProvider;
 import io.quarkus.vertx.http.runtime.CurrentVertxRequest;
 import io.quarkus.vertx.http.runtime.HttpBuildTimeConfig;
 import io.quarkus.vertx.http.runtime.HttpConfiguration;
+import io.quarkus.vertx.http.runtime.HttpHostConfigSource;
 import io.quarkus.vertx.http.runtime.RouterProducer;
 import io.quarkus.vertx.http.runtime.VertxHttpRecorder;
 import io.quarkus.vertx.http.runtime.attribute.ExchangeAttributeBuilder;
@@ -179,6 +182,13 @@ class VertxHttpProcessor {
                 shutdownConfig, executorBuildItem.getExecutorProxy());
 
         return new ServiceStartBuildItem("vertx-http");
+    }
+
+    @BuildStep
+    void hostDefault(BuildProducer<RunTimeConfigurationSourceBuildItem> serviceProviderBuildItem) {
+        serviceProviderBuildItem
+                .produce(new RunTimeConfigurationSourceBuildItem(HttpHostConfigSource.class.getName(),
+                        OptionalInt.of(Integer.MIN_VALUE)));
     }
 
     @Record(ExecutionTime.RUNTIME_INIT)

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpConfiguration.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpConfiguration.java
@@ -34,8 +34,14 @@ public class HttpConfiguration {
 
     /**
      * The HTTP host
+     *
+     * In dev/test mode this defaults to localhost, in prod mode this defaults to 0.0.0.0
+     *
+     * Defaulting to 0.0.0.0 makes it easier to deploy Quarkus to container, however it
+     * is not suitable for dev/test mode as other people on the network can connect to your
+     * development machine.
      */
-    @ConfigItem(defaultValue = "0.0.0.0")
+    @ConfigItem
     public String host;
 
     /**

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpHostConfigSource.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpHostConfigSource.java
@@ -1,0 +1,49 @@
+package io.quarkus.vertx.http.runtime;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Map;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import io.quarkus.runtime.LaunchMode;
+
+/**
+ * Sets the default host config value, depending on the launch mode.
+ * <p>
+ * This can't be done with a normal default, is it changes based on the mode,
+ * so instead it is provided as a low priority config source.
+ */
+public class HttpHostConfigSource implements ConfigSource, Serializable {
+
+    public static final String QUARKUS_HTTP_HOST = "quarkus.http.host";
+
+    private final int priority;
+
+    public HttpHostConfigSource(int priority) {
+        this.priority = priority;
+    }
+
+    @Override
+    public Map<String, String> getProperties() {
+        return Collections.singletonMap(QUARKUS_HTTP_HOST, getValue(QUARKUS_HTTP_HOST));
+    }
+
+    @Override
+    public int getOrdinal() {
+        return priority;
+    }
+
+    @Override
+    public String getValue(String propertyName) {
+        if (propertyName.equals(QUARKUS_HTTP_HOST)) {
+            return LaunchMode.current().isDevOrTest() ? "localhost" : "0.0.0.0";
+        }
+        return null;
+    }
+
+    @Override
+    public String getName() {
+        return "Quarkus HTTP Host Default Value";
+    }
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -187,10 +187,15 @@ public class VertxHttpRecorder {
             ConfigInstantiator.handleObject(buildConfig);
             HttpConfiguration config = new HttpConfiguration();
             ConfigInstantiator.handleObject(config);
+            if (config.host == null) {
+                //HttpHostConfigSource does not come into play here
+                config.host = "localhost";
+            }
             Router router = Router.router(vertx);
             if (hotReplacementHandler != null) {
                 router.route().order(Integer.MIN_VALUE).blockingHandler(hotReplacementHandler);
             }
+
             Handler<HttpServerRequest> root = router;
             LiveReloadConfig liveReloadConfig = new LiveReloadConfig();
             ConfigInstantiator.handleObject(liveReloadConfig);

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/JaxRSTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/JaxRSTestCase.java
@@ -16,6 +16,7 @@ import org.junit.jupiter.api.Test;
 
 import io.quarkus.it.rest.TestResource;
 import io.quarkus.test.common.http.TestHTTPEndpoint;
+import io.quarkus.test.junit.DisabledOnNativeImage;
 import io.quarkus.test.junit.QuarkusTest;
 import io.restassured.RestAssured;
 import io.restassured.parsing.Parser;
@@ -35,6 +36,7 @@ public class JaxRSTestCase {
     }
 
     @Test
+    @DisabledOnNativeImage //the native image tests will bind to 0.0.0.0 as the image is a production image, but the test datasource in the JVM will want to use localhost
     public void testConfigInjectionOfPort() {
         String host = ConfigProvider.getConfig().getOptionalValue("quarkus.http.host", String.class).orElse("0.0.0.0");
         RestAssured.when().get("/test/config/host").then().body(is(host));

--- a/test-framework/common/src/main/java/io/quarkus/test/common/NativeImageLauncher.java
+++ b/test-framework/common/src/main/java/io/quarkus/test/common/NativeImageLauncher.java
@@ -43,7 +43,7 @@ public class NativeImageLauncher implements Closeable {
     private final Class<?> testClass;
     private final String profile;
     private Process quarkusProcess;
-    private final int port;
+    private int port;
     private final int httpsPort;
     private final long imageWaitTime;
     private final Map<String, String> systemProps = new HashMap<>();
@@ -142,6 +142,7 @@ public class NativeImageLauncher implements Closeable {
 
             System.setProperty("quarkus.http.port", portCapturingProcessReader.port.toString()); //set the port as a system property in order to have it applied to Config
             System.setProperty("quarkus.http.test-port", portCapturingProcessReader.port.toString()); // needed for RestAssuredManager
+            port = portCapturingProcessReader.port;
             installAndGetSomeConfig(); // reinitialize the configuration to make sure the actual port is used
             System.setProperty("test.url", TestHTTPResourceManager.getUri());
         } else {
@@ -365,7 +366,9 @@ public class NativeImageLauncher implements Closeable {
 
         @Override
         protected void handleError(IOException e) {
-            portDetermined(null);
+            if (!portDetermined) {
+                portDetermined(null);
+            }
         }
 
         public void awaitForPort() throws InterruptedException {

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java
@@ -48,6 +48,8 @@ import io.quarkus.deployment.dev.DevModeContext;
 import io.quarkus.deployment.dev.DevModeMain;
 import io.quarkus.deployment.util.FileUtil;
 import io.quarkus.dev.appstate.ApplicationStateNotification;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ProfileManager;
 import io.quarkus.runtime.util.ClassPathUtils;
 import io.quarkus.test.common.PathTestHelper;
 import io.quarkus.test.common.PropertyTestUtil;
@@ -147,6 +149,8 @@ public class QuarkusDevModeTest
 
     @Override
     public void beforeAll(ExtensionContext context) throws Exception {
+        //set the right launch mode in the outer CL, used by the HTTP host config source
+        ProfileManager.setLaunchMode(LaunchMode.DEVELOPMENT);
         originalRootLoggerHandlers = rootLogger.getHandlers();
         rootLogger.addHandler(inMemoryLogHandler);
     }

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java
@@ -58,6 +58,8 @@ import io.quarkus.builder.BuildStep;
 import io.quarkus.builder.item.BuildItem;
 import io.quarkus.deployment.util.FileUtil;
 import io.quarkus.runner.bootstrap.AugmentActionImpl;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ProfileManager;
 import io.quarkus.test.common.PathTestHelper;
 import io.quarkus.test.common.PropertyTestUtil;
 import io.quarkus.test.common.RestAssuredURLManager;
@@ -338,6 +340,8 @@ public class QuarkusUnitTest
 
     @Override
     public void beforeAll(ExtensionContext extensionContext) throws Exception {
+        //set the right launch mode in the outer CL, used by the HTTP host config source
+        ProfileManager.setLaunchMode(LaunchMode.TEST);
         if (beforeAllCustomizer != null) {
             beforeAllCustomizer.run();
         }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -69,6 +69,7 @@ import io.quarkus.builder.BuildStep;
 import io.quarkus.deployment.builditem.TestAnnotationBuildItem;
 import io.quarkus.deployment.builditem.TestClassBeanBuildItem;
 import io.quarkus.deployment.builditem.TestClassPredicateBuildItem;
+import io.quarkus.runtime.LaunchMode;
 import io.quarkus.runtime.configuration.ProfileManager;
 import io.quarkus.runtime.test.TestHttpEndpointProvider;
 import io.quarkus.test.common.PathTestHelper;
@@ -568,6 +569,8 @@ public class QuarkusTestExtension
 
     @Override
     public void beforeAll(ExtensionContext context) throws Exception {
+        //set the right launch mode in the outer CL, used by the HTTP host config source
+        ProfileManager.setLaunchMode(LaunchMode.TEST);
         if (isNativeTest(context)) {
             return;
         }

